### PR TITLE
Adding OSC TripleO Heat Template (THT) and Kolla up stream design documents

### PR DESCRIPTION
**Pull request by emanoelxavier**
This PR adds the first drafts for the OSC TripleO and Kolla up stream design documents. It covers:
#### Contains
* First draft of the OSC TripleO-Heat-Template (THT) 
* Highlights the template parameters for configuration, environment variables and mount volumes
* Placeholders for: TLS, HA, service ports
#### Next
@sudhirappaji I will work with you to ensure we are tracking the following items as user stories:
* Configuration for OSC single instance vs. HA
* TLS certs, secret keys and passwords
* Service configuration (i.e.: ports)

#### Contains
* High level design
* OSC container j2 file
#### Next
@sudhirappaji I will work with you to ensure we are tracking the following items as user stories:
* Evaluate different packaging options. If Kolla:
* Review design proposal with Kolla community and design remaining pieces

### CODE REVIEW

File: designs/tripleo/osc-tripleo.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
+# OSC TripleO
```
Comment (sudhirappaji) [inline designs/tripleo/osc-tripleo.md]: Can't we call the document as OSC Integration with TripleO? we could have different chapters like THT, Puppet (if needed), services etc. I see RH as documentation on partner integration.

Comment (sudhirappaji) [inline designs/tripleo/osc-tripleo.md]: what is the plan to add these templates? meaning which repository. I see ODL as separate repository for puppet etc in ODL github?

Comment (sudhirappaji) [reply designs/tripleo/osc-tripleo.md]: Do you plan to add a note how to work with TripleO community, to get approval for these templates?

File: designs/tripleo/osc-tripleo.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+The OSC THT will be  `/tripleo-heat-templates/docker/services/osc.yaml`. The resource must be registered in the environment file `tripleo-heat-templates/environments/docker.yaml` as follows:  
+```OS::TripleO::Services::OSC: ../docker/services/osc.yaml```  
+
+With these files in place the user can include the OSC service as a resource in the *overcloud* by adding `OS::TripleO::Services::OSC` to the *controller* role during deployment and providing *--environment* files with the values of the parameters defined on the template.   
+
+#### Template Content
+The file `../docker/services/osc.yaml` starts with the template version and a description, followed a set of parameters, a list of resources and outputs generated by the template. For details on THTs, see the [guide](#tht-guide).  
+
+```yaml  
+heat_template_version: rocky
```
Comment (sudhirappaji) [inline designs/tripleo/osc-tripleo.md]: I feel we need to target for queens release, for acceptance of these template and making OSC deployed using queens release. Once this is working, we can finally target for Rocky. The first step should be for queens'.

File: designs/tripleo/osc-tripleo.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+description: >
+  Open Security Controller: service for orchestration of virtual network security functions.
+
+ parameters:
+  #  The paramater EndpointMap is required for all service templates*.
+  EndpointMap:
```
Comment (sudhirappaji) [inline designs/tripleo/osc-tripleo.md]: Not clear how the EndpointMap contains and how is this used? I tried to google, but not much information is available.
could you add an example? and give description about this?

File: designs/tripleo/osc-tripleo.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+*[THT Tutorial](#tht-tutorial)  
+
+**[Kolla Container Repos](#kolla-container-repos)  
+
+> **Assumption:** The images in the Kolla Repos will be picked and replicated in the TripleO Repo https://hub.docker.com/r/tripleoupstream 
+
+#### OSC Environment Variables  
+Environment variables required by OSC can be provided by THT parameters as shown `OSC_ENV_VAR={get_param: OSCEnvVar}`.  
+TODO: Replace placeholder with actual environment variable Srequired by OSC.
```
Comment (sudhirappaji) [inline designs/tripleo/osc-tripleo.md]: typo Srequired.

File: designs/tripleo/osc-tripleo.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    type: string
+	
+resources:
+  # This resource containers a static list of configurations and scripts necessary 
+  # for the deployment of container services in the *overcloud*
+  ContainersCommon:
+    type: ./containers-common.yaml
+
+outputs:
+  # Service templates must output a role_data value containing the config_settings and step_config for the service configuration**
```
Comment (sudhirappaji) [inline designs/tripleo/osc-tripleo.md]: I do not understand? can we add description what is role data? or point to relevant link, which helps us to understand better?

File: designs/tripleo/osc-tripleo.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    description: Role data for the OSC service.
+    value:
+      config_settings:
+        docker_config:
+        step_1:
+          osc:
+            start_order: 0
+            image: &osc_image {get_param: OSCDockerImage}
+            privileged: false
+            net: host
```
Comment (sudhirappaji) [inline designs/tripleo/osc-tripleo.md]: could we have description of each of the keywords?

File: designs/tripleo/osc-tripleo.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+The intent of this work is to allow OSC to be more easily deployed and configured in a OpenStack cloud by using TripleO. OSC will be deployed in the **controller** node of the **overcloud** as shown below:  
+![](./images/osc-tripleo-services.png)  
+*OSC In the TripleO Overcloud*  
+
+## Constraints and Assumptions
+### OSC Packaged Binaries
+TripleO can only deploy OSC through a template if the OSC binaries are available in a public repository. There are multiple paths to achieve this:  
+1.  RPM package.
+2.  Docker container image on an arbitrary Docker repository.
+3.  Docker container image on a Kolla repository.
```
Comment (sudhirappaji) [inline designs/tripleo/osc-tripleo.md]: How is the Kolla Image downloaded as part of heat/puppet template configuration? as i understand, if it is RPM, they use 'package' keyword' to download and run the yum install ==> as part of puppet configuration, 
How is this for Kolla image download and installation in comparison to RPM? or would this be part of separate US?

File: designs/osc-packaging/kolla/osc-kolla.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
+# OSC Kolla
+[Kolla](#kolla-home) is an OpenStack project that provides production-ready containers and deployment tools for operating OpenStack clouds. This document outlines the contributions to be up streamed to Kolla so that the OSC image is present in the [Kolla container repository](#kolla-container-repos).
```
Comment (sudhirappaji) [inline designs/osc-packaging/kolla/osc-kolla.md]: Do we have any process to upload OSC kolla container image to kolla repository? how is this done? any procedure to follow to get OSC repository place-holder? do we need to work with the kolla community for putting the osc container image?

File: designs/tripleo/osc-tripleo.md (hunk 1 above)
Comment (emanoelxavier) [reply designs/tripleo/osc-tripleo.md]: The document mentions that here "The goal is for all the changes described on this section to be upstreamed to the [THT repository](#tripleo-repo)"  and "### [TripleO Repo](https://github.com/openstack/tripleo-heat-templates) . ODL was just an example we were looking at. ODL has changes on the same repo and also on the puppet-tripleo repo on GitHub. I am assuming we wont need the latter, just the repo https://github.com/openstack/tripleo-heat-templates. "How" we will work with the tripleo community is outside of design IMO we will definitely need to describe that through RFS (we already have), blueprints, etc. But the process is orthogonal to the design and needs to be discussed more with RH.

File: designs/tripleo/osc-tripleo.md (hunk 2 above)
Comment (emanoelxavier) [reply designs/tripleo/osc-tripleo.md]: We will need to change to rocky once we are ready to up stream but for now I agree. Changing to queens.

File: designs/tripleo/osc-tripleo.md (hunk 3 above)
Comment (emanoelxavier) [reply designs/tripleo/osc-tripleo.md]: It is not fully clear to me either. However I did put a reference to the document that says it is required "### [THT Guide](https://docs.openstack.org/heat/pike/template_guide/hot_guide.html) " . We can clarify that once we review with Luke. I am tracking another user story to update this document after the first review with RH and will ensure to add that detail. Adding a TODO now.

File: designs/tripleo/osc-tripleo.md (hunk 5 above)
Comment (emanoelxavier) [reply designs/tripleo/osc-tripleo.md]: This is also a required output for all TripleO services. It outputs value settings used to configure the services on the overcloud. The doc has a reference for that here, thus the `*` (was **, typo) in the end of the sentence. [THT Guide](https://docs.openstack.org/heat/pike/template_guide/hot_guide.html)

File: designs/tripleo/osc-tripleo.md (hunk 6 above)
Comment (emanoelxavier) [reply designs/tripleo/osc-tripleo.md]: These map directly to the parameters used for the docker run command for OSC. I have added a note for that and a link to the doc https://docs.docker.com/engine/reference/run/ which describes the parameters.

File: designs/tripleo/osc-tripleo.md (hunk 7 above)
Comment (emanoelxavier) [reply designs/tripleo/osc-tripleo.md]: Yes these details need to be investigated in another US, I have proposed 10 new ones :). In short TripleO infra clones/copies all the images in the Kolla repo (i.e.: https://hub.docker.com/r/kolla) to the tripleO repo (https://hub.docker.com/r/tripleoupstream/) which means that THT can reference those images directly when triggering docker run. If we do use RPM instead then yes likely we will need a separate puppet script to download and install the RPM.

File: designs/osc-packaging/kolla/osc-kolla.md (hunk 8 above)
Comment (emanoelxavier) [reply designs/osc-packaging/kolla/osc-kolla.md]: Yes to all those questions. I have proposed more user stories to track that. But before we involve the community etc we will need to determine whether we will use Kolla (there is also a user story for that). The main purpose of this initial work was to parse the ODL material, learn how OOO deals with basic configuration and identify the next steps/areas for investigation.

**Pull request merged**